### PR TITLE
Convert MySQL fields to 4 byte unicode 

### PR DESCRIPTION
**Pull request by dimaursu**
This will make Diaspora\* running on postgres and on mysql behave consistently: in the postgres version, you can use 4-byte UTF8, or full UTF8. In MySQL `utf8` == the 3 byte subset of UTF8, while `utf8mb4` is the true, full UTF8 encoding.

### CODE REVIEW

File: db/schema.rb [hunk 1]
```diff
@@ -429,11 +433,11 @@
...
   create_table "reports", force: true do |t|
     t.integer  "item_id",                    null: false
-    t.string   "item_type",                  null: false
     t.boolean  "reviewed",   default: false
     t.text     "text"
     t.datetime "created_at"
     t.datetime "updated_at"
+    t.string   "item_type",                  null: false
     t.integer  "user_id",                    null: false
   end
```
Comment (jhass) [inline db/schema.rb]: No actual schema changes in the migration, please don't commit schema.rb changes.

File: db/migrate/20150106050733_set_mysql_to_unicode_mb4.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  UTF8_PAIRS = {
+    'comments': 'text',
+    'messages': 'text',
+    'poll_answers': 'answer',
+    'polls': 'question',
+    'posts': 'text',
+  }
+
+  def self.up
+    if ENV['DB'] == 'mysql'
```
Comment (jhass) [inline db/migrate/20150106050733_set_mysql_to_unicode_mb4.rb]: I plan to eventually get rid of the DB environment variable, I think you should be able to query the adapter through AR::Base.config or something like that.

**Comment by jhass** [comment]
I guess you want to add to the big fat changelog entries here, so podmins notice to update their database.yml.

File: db/schema.rb [hunk 3]
```diff
@@ -11,7 +11,10 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema.define(version: 20141216213423) do
+ActiveRecord::Schema.define(version: 20150106050733) do
```
Comment (jhass) [inline db/schema.rb]: That is, the version change should be committed of course.

File: db/migrate/20150106050733_set_mysql_to_unicode_mb4.rb (hunk 2 above)
Comment (jhass) [inline db/migrate/20150106050733_set_mysql_to_unicode_mb4.rb]: Actually I just remembered we already have https://github.com/diaspora/diaspora/blob/develop/lib/configuration_methods.rb#L121 which can be called through `AppConfig.postgres?`. Don't need that often luckily!

File: Gemfile [hunk 4]
```diff
@@ -57,7 +57,7 @@ gem 'sass-rails',     '4.0.4'
 
 # Database
 
-ENV['DB'] ||= 'mysql'
+ENV['DB'] ||= 'postgres'
```
Comment (jhass) [inline Gemfile]: As much as I'd wish for, but nope...

File: Gemfile [hunk 5]
```diff
@@ -88,6 +88,7 @@ gem 'rails-assets-jquery',   '1.11.1' # Should be kept in sync with jquery-rails
 gem 'js_image_paths',        '0.0.1'
 gem 'js-routes',             '0.9.9'
 gem 'rails-assets-punycode', '1.3.2'
+gem 'rails-assets-twemoji'
```
Comment (jhass) [inline Gemfile]: Let's keep that change separate, since I disagree with it while agreeing about changing the database encoding.

**Comment by goobertron** [comment]
If this is merged, please can we also have a means of blocking these emoji from view?

**Comment by jhass** [comment]
@goobertron it's actually not graphic emojis for this one and actually just badly argued through the unicode emojis, so it's rather unreleated to them, they're just a side effect. There's strong technical reasons to do this.

**Comment by goobertron** [comment]
Oh right, I was basing my comment on the opening post, which includes graphics. I guess that's Github's fault. I'd still like an opt-out from seeing any emoji, however they are displayed.

**Comment by jhass** [comment]
For this one, just remove any fonts that would display them... :P

File: db/migrate/20150106050733_set_mysql_to_unicode_mb4.rb [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    ActiveRecord::Base.connection.columns(table).each do |column|
+      # build a hash with all the columns that contain text
+      if (column.type == :string) || (column.type == :text)
+        UTF8_PAIRS[table] = { 'name' => column.name, 'type' => column.sql_type }
+      end
+    end
+  end
+
+  def self.up
+    change_encoding('utf8mb4') if AppConfig.mysql?
```
Comment (dimaursu) [inline db/migrate/20150106050733_set_mysql_to_unicode_mb4.rb]: @jhass I'ts beautiful!

File: db/migrate/20150106050733_set_mysql_to_unicode_mb4.rb [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+  # Converts the tables and strings columns to utf8mb4, which is the true, full
+  # unicode support in MySQl
+  UTF8_PAIRS = {}
+  tables = ActiveRecord::Base.connection.tables
+
+  tables.each do |table|
+    ActiveRecord::Base.connection.columns(table).each do |column|
+      # build a hash with all the columns that contain text
+      if (column.type == :string) || (column.type == :text)
+        UTF8_PAIRS[table] = { 'name' => column.name, 'type' => column.sql_type }
```
Comment (jhass) [inline db/migrate/20150106050733_set_mysql_to_unicode_mb4.rb]: I'd use symbol keys here :)

**Comment by dimaursu** [comment]
@jhass `ROW_FORMAT=DYNAMIC`, in combination with InnoDB engine, and those 3 parameters above (now set in the migration, no need to manually touch configs), increases the maximum size of the index columns from 767 bytes to 3072 (something like that); that gives us 768 chars for the indexes, in the worst case scenario (a whole index row is made out of emojis would be a good example). MySQL only considers this scenario, that's why it spits out errors.

**Comment by jhass** [comment]
Okay, I see now how that transforms existing databases. How do we get to the same state when doing `rake db:create db:schema:load`?

**Comment by dimaursu** [comment]
No clue about that. But if you give me an idea or some pointers, I will gladly add that too.

File: db/migrate/20150106050733_set_mysql_to_unicode_mb4.rb [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  # Converts the tables and strings columns to utf8mb4, which is the true, full
+  # unicode support in MySQl
+  UTF8_PAIRS = {}
+  tables = ActiveRecord::Base.connection.tables
+
+  tables.each do |table|
+    ActiveRecord::Base.connection.columns(table).each do |column|
+      # build a hash with all the columns that contain text
+      if (column.type == :string) || (column.type == :text)
+        UTF8_PAIRS[table] = { :name => column.name, :type => column.sql_type }
```
Comment (dimaursu) [inline db/migrate/20150106050733_set_mysql_to_unicode_mb4.rb]: This is not good. We should have the key the column. This is a serious error from my side, sorry.

File: db/schema.rb [hunk 9]
```diff
@@ -31,19 +31,19 @@
...
-    t.integer  "shareable_id",                    null: false
-    t.integer  "aspect_id",                       null: false
-    t.datetime "created_at",                      null: false
-    t.datetime "updated_at",                      null: false
-    t.string   "shareable_type", default: "Post", null: false
+    t.integer  "shareable_id",   null: false
+    t.integer  "aspect_id",      null: false
+    t.datetime "created_at",     null: false
+    t.datetime "updated_at",     null: false
+    t.string   "shareable_type"
```
Comment (jhass) [inline db/schema.rb]: Where did the default went?

File: db/schema.rb [hunk 10]
```diff
@@ -31,19 +31,19 @@
...
 
   add_index "aspect_visibilities", ["aspect_id"], name: "index_aspect_visibilities_on_aspect_id", using: :btree
-  add_index "aspect_visibilities", ["shareable_id", "shareable_type", "aspect_id"], name: "shareable_and_aspect_id", using: :btree
-  add_index "aspect_visibilities", ["shareable_id", "shareable_type"], name: "index_aspect_visibilities_on_shareable_id_and_shareable_type", using: :btree
+  add_index "aspect_visibilities", ["shareable_id", "shareable_type", "aspect_id"], name: "shareable_and_aspect_id", length: {"shareable_id"=>nil, "shareable_type"=>189, "aspect_id"=>nil}, using: :btree
+  add_index "aspect_visibilities", ["shareable_id", "shareable_type"], name: "index_aspect_visibilities_on_shareable_id_and_shareable_type", length: {"shareable_id"=>nil, "shareable_type"=>190}, using: :btree
 
   create_table "aspects", force: true do |t|
-    t.string   "name",                             null: false
+    t.string   "name"
```
Comment (jhass) [inline db/schema.rb]: Where did the not null constraint went?

File: db/schema.rb [hunk 11]
```diff
@@ -61,47 +61,47 @@
   end
 
   create_table "chat_contacts", force: true do |t|
-    t.integer "user_id",                  null: false
-    t.string  "jid",                      null: false
+    t.integer "user_id",                   null: false
+    t.string  "jid",          limit: 3071
```
Comment (jhass) [inline db/schema.rb]: Left over/typo?

File: db/schema.rb [hunk 12]
```diff
@@ -61,47 +61,47 @@
...
     t.string  "name"
     t.string  "ask",          limit: 128
-    t.string  "subscription", limit: 128, null: false
+    t.string  "subscription", limit: 128
   end
 
-  add_index "chat_contacts", ["user_id", "jid"], name: "index_chat_contacts_on_user_id_and_jid", unique: true, using: :btree
+  add_index "chat_contacts", ["user_id", "jid"], name: "index_chat_contacts_on_user_id_and_jid", length: {"user_id"=>nil, "jid"=>190}, using: :btree
```
Comment (jhass) [inline db/schema.rb]: uniqueness constraint went missing.

File: db/schema.rb [hunk 13]
```diff
@@ -61,47 +61,47 @@
...
     t.text     "author_signature"
     t.text     "parent_author_signature"
-    t.datetime "created_at",                                          null: false
-    t.datetime "updated_at",                                          null: false
-    t.integer  "likes_count",                        default: 0,      null: false
-    t.string   "commentable_type",        limit: 60, default: "Post", null: false
+    t.datetime "created_at",                                     null: false
+    t.datetime "updated_at",                                     null: false
+    t.integer  "likes_count",                        default: 0, null: false
+    t.string   "commentable_type",        limit: 60
```
Comment (jhass) [inline db/schema.rb]: default, not null

File: db/schema.rb [hunk 14]
```diff
@@ -366,7 +366,7 @@
     t.string   "provider_display_name"
     t.string   "actor_url"
     t.string   "objectId"
-    t.string   "root_guid"
+    t.string   "root_guid",             limit: 64
```
Comment (jhass) [inline db/schema.rb]: New limit not present in migration.

File: db/schema.rb [hunk 15]
```diff
@@ -376,18 +376,18 @@
     t.string   "frame_name"
     t.boolean  "favorite",                         default: false
     t.string   "facebook_id"
-    t.string   "tweet_id"
+    t.string   "tweet_id",              limit: 64
```
Comment (jhass) [inline db/schema.rb]: New limit

File: db/schema.rb [hunk 16]
```diff
@@ -457,16 +457,16 @@
...
-    t.datetime "updated_at",                                 null: false
-    t.boolean  "hidden",                    default: false,  null: false
-    t.integer  "contact_id",                                 null: false
-    t.string   "shareable_type", limit: 60, default: "Post", null: false
+    t.integer  "shareable_id",                              null: false
+    t.datetime "created_at",                                null: false
+    t.datetime "updated_at",                                null: false
+    t.boolean  "hidden",                    default: false, null: false
+    t.integer  "contact_id",                                null: false
+    t.string   "shareable_type", limit: 60
```
Comment (jhass) [inline db/schema.rb]: default, not null

File: db/schema.rb [hunk 17]
```diff
@@ -529,8 +529,8 @@
     t.boolean  "getting_started",                                default: true,  null: false
     t.boolean  "disable_mail",                                   default: false, null: false
     t.string   "language"
-    t.string   "email",                                          default: "",    null: false
-    t.string   "encrypted_password",                             default: "",    null: false
+    t.string   "email"
+    t.string   "encrypted_password"
```
Comment (jhass) [inline db/schema.rb]: default, not null

File: db/schema.rb [hunk 18]
```diff
@@ -561,10 +561,10 @@
   end
 
   add_index "users", ["authentication_token"], name: "index_users_on_authentication_token", unique: true, using: :btree
-  add_index "users", ["email"], name: "index_users_on_email", using: :btree
-  add_index "users", ["invitation_service", "invitation_identifier"], name: "index_users_on_invitation_service_and_invitation_identifier", unique: true, using: :btree
+  add_index "users", ["email"], name: "index_users_on_email", length: {"email"=>191}, using: :btree
+  add_index "users", ["invitation_service", "invitation_identifier"], name: "index_users_on_invitation_service_and_invitation_identifier", unique: true, length: {"invitation_service"=>64, "invitation_identifier"=>nil}, using: :btree
```
Comment (jhass) [inline db/schema.rb]: The nil is intentional?

File: Changelog.md [hunk 19]
```diff
@@ -32,6 +32,10 @@ series and run our comprehensive test suite against it.
 ## Change in defaults.yml
 The default for including jQuery from a CDN has changed. If you want to continue to include it from a CDN, please explicitly set the `jquery_cdn` setting to `true` in diaspora.yml.
 
+## Change in database.yml
+For MySQL databases, replace `charset: utf8` with `encoding: utf8mb4`. This is enables full UTF8 support (4bytes characters), including standard emoji characters. See `database.yml.example` for reference.
+Also, do not forget to remove `collation: utf8_bin`. It will choose a compatible one automatically.
```
Comment (Flaburgan) [inline Changelog.md]: This is not precise enough imo. Where is that setting? Which file, which line? If you really need a podmin action, you should explain precisely what he should do.

File: Changelog.md [hunk 20]
```diff
@@ -32,6 +32,10 @@ series and run our comprehensive test suite against it.
 ## Change in defaults.yml
 The default for including jQuery from a CDN has changed. If you want to continue to include it from a CDN, please explicitly set the `jquery_cdn` setting to `true` in diaspora.yml.
 
+## Change in database.yml
+For MySQL databases, replace `charset: utf8` with `encoding: utf8mb4` in the file `config/database.yml`. This is enables full UTF8 support (4bytes characters), including standard emoji characters. See `database.yml.example` for reference.
+Also, do not forget to remove `collation: utf8_bin`. It will choose a compatible one automatically.
```
Comment (dimaursu) [inline Changelog.md]: With `database.yml.example` at hand, I guess any podmin will figure this out.

**Comment by jhass** [comment]
https://travis-ci.org/diaspora/diaspora/jobs/46997679#L357

**Comment by denschub** [comment]
As promised, ran the migration on a clone of my production database.

Size on disk before: 43062 MB
Size on disk after: 43167 MB

Full migration: 304.7380s, full log below:

```
== 20150106050733 SetMysqlToUnicodeMb4: migrating =============================
-- remove_index("aspect_visibilities", {:name=>"shareable_and_aspect_id"})
   -> 0.0574s
-- add_index("aspect_visibilities", ["shareable_id", "shareable_type", "aspect_id"], {:name=>"shareable_and_aspect_id", :length=>{"shareable_type"=>189}, :using=>:btree})
   -> 16.0878s
-- remove_index("aspect_visibilities", {:name=>"index_aspect_visibilities_on_shareable_id_and_shareable_type"})
   -> 0.0500s
-- add_index("aspect_visibilities", ["shareable_id", "shareable_type"], {:name=>"index_aspect_visibilities_on_shareable_id_and_shareable_type", :length=>{"shareable_type"=>190}, :using=>:btree})
   -> 9.1848s
-- remove_index("chat_contacts", {:name=>"index_chat_contacts_on_user_id_and_jid"})
   -> 0.1278s
-- add_index("chat_contacts", ["user_id", "jid"], {:name=>"index_chat_contacts_on_user_id_and_jid", :length=>{"jid"=>190}, :using=>:btree, :unique=>true})
   -> 0.1719s
-- remove_index("comments", {:name=>"index_comments_on_guid"})
   -> 0.2913s
-- add_index("comments", ["guid"], {:name=>"index_comments_on_guid", :length=>{"guid"=>191}, :using=>:btree, :unique=>true})
   -> 33.2449s
-- remove_index("likes", {:name=>"index_likes_on_guid"})
   -> 0.3831s
-- add_index("likes", ["guid"], {:name=>"index_likes_on_guid", :length=>{"guid"=>191}, :using=>:btree, :unique=>true})
   -> 70.5352s
-- remove_index("o_embed_caches", {:name=>"index_o_embed_caches_on_url"})
   -> 0.0831s
-- add_index("o_embed_caches", ["url"], {:name=>"index_o_embed_caches_on_url", :length=>{"url"=>191}, :using=>:btree})
   -> 2.0182s
-- remove_index("participations", {:name=>"index_participations_on_guid"})
   -> 0.1053s
-- add_index("participations", ["guid"], {:name=>"index_participations_on_guid", :length=>{"guid"=>191}, :using=>:btree})
   -> 96.0637s
-- remove_index("people", {:name=>"index_people_on_diaspora_handle"})
   -> 0.0838s
-- add_index("people", ["diaspora_handle"], {:name=>"index_people_on_diaspora_handle", :unique=>true, :length=>{"diaspora_handle"=>191}})
   -> 0.9094s
-- remove_index("people", {:name=>"index_people_on_guid"})
   -> 0.0497s
-- add_index("people", ["guid"], {:name=>"index_people_on_guid", :length=>{"guid"=>191}, :using=>:btree, :unique=>true})
   -> 0.3075s
-- remove_index("photos", {:name=>"index_photos_on_status_message_guid"})
   -> 0.0415s
-- add_index("photos", ["status_message_guid"], {:name=>"index_photos_on_status_message_guid", :length=>{"status_message_guid"=>191}, :using=>:btree})
   -> 2.8110s
-- remove_index("posts", {:name=>"index_posts_on_guid"})
   -> 0.0748s
-- add_index("posts", ["guid"], {:name=>"index_posts_on_guid", :length=>{"guid"=>191}, :using=>:btree, :unique=>true})
   -> 26.3498s
-- remove_index("posts", {:name=>"index_posts_on_status_message_guid_and_pending"})
   -> 0.1532s
-- add_index("posts", ["status_message_guid", "pending"], {:name=>"index_posts_on_status_message_guid_and_pending", :length=>{"status_message_guid"=>190}, :using=>:btree})
   -> 6.0173s
-- remove_index("posts", {:name=>"index_posts_on_status_message_guid"})
   -> 0.0775s
-- add_index("posts", ["status_message_guid"], {:name=>"index_posts_on_status_message_guid", :length=>{"status_message_guid"=>191}, :using=>:btree})
   -> 6.4922s
-- remove_index("rails_admin_histories", {:name=>"index_rails_admin_histories"})
   -> 0.0716s
-- add_index("rails_admin_histories", ["item", "table", "month", "year"], {:name=>"index_rails_admin_histories", :length=>{"table"=>188}, :using=>:btree})
   -> 0.1415s
-- remove_index("schema_migrations", {:name=>"unique_schema_migrations"})
   -> 0.7815s
-- add_index("schema_migrations", ["version"], {:name=>"unique_schema_migrations", :length=>{"version"=>191}, :using=>:btree})
   -> 0.2091s
-- remove_index("services", {:name=>"index_services_on_type_and_uid"})
   -> 0.0873s
-- add_index("services", ["type", "uid"], {:name=>"index_services_on_type_and_uid", :length=>{"type"=>64, "uid"=>127}, :using=>:btree})
   -> 0.2988s
-- remove_index("taggings", {:name=>"index_taggings_on_taggable_id_and_taggable_type_and_context"})
   -> 0.0946s
-- add_index("taggings", ["taggable_id", "taggable_type", "context"], {:name=>"index_taggings_on_taggable_id_and_taggable_type_and_context", :length=>{"taggable_type"=>95, "context"=>95}, :using=>:btree})
   -> 23.9894s
-- remove_index("tags", {:name=>"index_tags_on_name"})
   -> 0.0702s
-- add_index("tags", ["name"], {:name=>"index_tags_on_name", :length=>{"name"=>191}, :using=>:btree, :unique=>true})
   -> 1.5240s
-- remove_index("users", {:name=>"index_users_on_invitation_service_and_invitation_identifier"})
   -> 0.1581s
-- add_index("users", ["invitation_service", "invitation_identifier"], {:name=>"index_users_on_invitation_service_and_invitation_identifier", :length=>{"invitation_service"=>64, "invitation_identifier"=>127}, :using=>:btree, :unique=>true})
   -> 0.5600s
-- remove_index("users", {:name=>"index_users_on_username"})
   -> 0.1131s
-- add_index("users", ["username"], {:name=>"index_users_on_username", :length=>{"username"=>191}, :using=>:btree, :unique=>true})
   -> 0.2943s
-- remove_index("users", {:name=>"index_users_on_email"})
   -> 0.1425s
-- add_index("users", ["email"], {:name=>"index_users_on_email", :length=>{"email"=>191}, :using=>:btree})
   -> 0.1858s
-- remove_index("notifications", {:name=>"index_notifications_on_target_type_and_target_id"})
   -> 0.1081s
-- add_index("notifications", ["target_type", "target_id"], {:name=>"index_notifications_on_target_type_and_target_id", :length=>{"target_type"=>190}, :using=>:btree})
   -> 4.1315s
== 20150106050733 SetMysqlToUnicodeMb4: migrated (304.7380s) ==================
```

:+1:

File: db/schema.rb (hunk 9 above)
Comment (jhass) [inline db/schema.rb]: This is still an issue.

File: db/schema.rb (hunk 10 above)
Comment (jhass) [inline db/schema.rb]: This too is still an issue.

File: db/schema.rb [hunk 21]
```diff
@@ -155,7 +155,7 @@
     t.string   "service"
     t.string   "identifier"
     t.boolean  "admin",        default: false
-    t.string   "language",     default: "en"
+    t.string   "language"
```
Comment (jhass) [inline db/schema.rb]: This one is making the tests fail, which doesn't mean the other ones shouldn't be fixed too.

File: db/schema.rb (hunk 14 above)
Comment (jhass) [inline db/schema.rb]: Still an issue.

**Comment by jhass** [comment]
Alright, let's risk this :P

Thank you!

**Comment by Flaburgan** [comment]
I run that on diaspora-fr and have http://paste.mrzyx.de/pxfy0zy2m

**Pull request merged**